raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -990,9 +1010,25 @@ private boolean handleFetchResponse(
                 OffsetAndEpoch nextFetchOffsetAndEpoch = new OffsetAndEpoch(
                     nextOffsetAndEpoch.nextFetchOffset(), nextOffsetAndEpoch.nextFetchOffsetEpoch());
 
+                state.highWatermark().ifPresent(highWatermark -> {
+                    if (nextFetchOffsetAndEpoch.offset < highWatermark.offset) {
+                        throw new KafkaException("The leader requested truncation to offset " +
+                            nextFetchOffsetAndEpoch.offset + ", which is below the current high watermark" +
+                            " " + highWatermark);
+                    }
+                });
+
                 log.truncateToEndOffset(nextFetchOffsetAndEpoch).ifPresent(truncationOffset -> {
                     logger.info("Truncated to offset {} after out of range error from leader {}",
                         truncationOffset, quorum.leaderIdOrNil());
+
+                    // Since the end offset has been updated, we should complete any delayed
+                    // reads at the end offset.
+
+                    // FIXME: Come up with a better solution for completing all
+                    fetchPurgatory.maybeComplete(
```

Comment:
At this moment the broker should know it's not the leader and hence its fetch purgatory should always be empty right?